### DISCUSSION TITLE
FreeSql接入Abp

### DISCUSSION THREAD

**Discussion by chenyigg** [dotnetcore/FreeSql#1274] [Q&A]
使用FreeSql作为Abp的ORM的话，是不是会让Abp原来的一些默认功能无法兼容？类似于权限部分，需要依赖EFCore的包。

**Comment by 2881099** [comment]
abp 接入，建议不要替换，可以考虑共存。事务用 efcore，查询用 freesql。

freesql 支持像 dapper 那样的用法，文档《你不知道的功能》有介绍，由于 freesql 可以移动识别 efcore 实体注解，所以这种接入成本最低。

**Reply by chenyigg** [reply]
我这边有点特殊，数据库由于是达梦，不支持efcore6.0，只能采取freesql

**Comment by 2881099** [comment]
# Ado.Net 扩展方法

提供了类似 Dapper 的使用方法，FreeSql 增加了 IDbConnection/IDbTransaction 对象的扩展方法 Select/Insert/Update/Delete 实现 CRUD。

```csharp
using FreeSql;

using (var conn = new SqlConnection(...))
{
  //IFreeSql fsql = conn.GetIFreeSql();
  //fsql.CodeFirst.IsNoneCommandParameter = true;
  //fsql.CodeFirst.IsSyncStructureToUpper = true;
  //fsql.Aop.CommandBefore += (_, e) => 
  //  Trace.WriteLine(e.Command.CommandText);
  //以上整个程序只需要设置一次

  conn.Select<T>().Where(...).ToList();

  conn.Insert(new T {}).ExecuteAffrows();
  conn.Update().SetSource(new T {}).ExecuteAffrows();
  conn.InsertOrUpdate().SetSource(new T {}).ExecuteAffrows();

  conn.Delete<T>().Where(...).ExecuteAffrows();
}
```

- 每个 SqlConnection GetFreeSql() 返回的 IFreeSql 实例相同；
- 可以对 fsql 设置 Aop 事件，比如监视 SQL；
- IFreeSql 自身的成员 IDbFirst、Transaction 不可用；

利用本功能可以快速将 FreeSql 使用到项目中，只需要处理好实体类的特性。

提示：FreeSql 兼容 EFCore 99% 的实体特性

**Comment by luoyunchong** [comment]
自然依旧要依赖EFCore，除非完全重新他的各个EFCore的模块，工作量比较大。


可参考这二个issues

https://github.com/dotnetcore/FreeSql/issues/712

https://github.com/dotnetcore/FreeSql/issues/593

**Reply by chenyigg** [reply]
好的，感谢
